gms/gossiper.cc

```diff
@@ -2150,27 +2275,38 @@ future<> gossiper::wait_for_gossip(std::chrono::milliseconds initial_delay, std:
 }
 
 future<> gossiper::wait_for_gossip_to_settle() {
-    static constexpr std::chrono::milliseconds GOSSIP_SETTLE_MIN_WAIT_MS{5000};
-    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
+    logger.info("Waiting for gossip to settle with query...");
     auto do_enable_features = [this] {
         return async([this] {
             if (!std::exchange(_gossip_settled, true)) {
                maybe_enable_features();
             }
         });
     };
+    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
     if (force_after == 0) {
         return do_enable_features();
     }
-    return wait_for_gossip(GOSSIP_SETTLE_MIN_WAIT_MS, force_after).then([this, do_enable_features] {
+    return wait_for_cluster_ring_status_to_match(*this).then([do_enable_features] {
```

Comment:
We should still respect `GOSSIP_SETTLE_MIN_WAIT_MS` and wait at least that long, which was there to let failure detector mark nodes as UP or DOWN appropriately (so that storage_proxy routes accordingly when servers start to accept connections).

Or better, we should mark all nodes as UP manually if we're returning from wait_for_cluster_ring_status_to_match() sooner.

Reply:
Why do we need the GOSSIP_SETTLE_MIN_WAIT_MS unconditionally? The node is marked as DOWN by default. Once we have a gossip update from other nodes, it will be marked as UP. 

So at the moment, the hash matches, information that is UP or DOWN is already update to date.

What do you mean by make it UP manually?

Reply:
> So at the moment, the hash matches, information that is UP or DOWN is already update to date.

`wait_for_cluster_ring_status_to_match()` doesn't wait for gossip updates in case all hashes match. So it races with ` gossiper::mark_alive()` and gossip echos. Can it return while some nodes are still not marked as UP?

> What do you mean by make it UP manually?

Call `real_mark_alive()` for the nodes which responded.

Reply:
A node is UP or DOWN is a per node view. We do not require all nodes have the same view at any point in time. It is not the job of wait_for_cluster_ring_status_to_match. 

Waiting for all pending real_mark_alive is needed only before we start the cql server in main.cc.

Perhaps,  we can wait for _pending_mark_alive_endpoints to be empty at the end of gossiper::wait_for_gossip_to_settle().

Reply:
> A node is UP or DOWN is a per node view. We do not require all nodes have the same view at any point in time. It is not the job of wait_for_cluster_ring_status_to_match.

Yes. I never suggested otherwise.


> 
> Waiting for all pending real_mark_alive is needed only before we start the cql server in main.cc.

Yes, but it's part of the responsibility of `wait_for_gossip_to_settle()` to wait for this. We should keep that responsibility. That was the reason we had GOSSIP_SETTLE_MIN_WAIT_MS.

> 
> Perhaps, we can wait for _pending_mark_alive_endpoints to be empty at the end of gossiper::wait_for_gossip_to_settle().

That seems less reliable since you assume that all mark alives are started by that time, unless we also wait for a gossip round.

Reply:
We have too places calling wait_for_gossip_to_settle. The one inside storage_service.cc does not really need nodes to be marked as up. 

```
 443     // Wait for gossip to settle so that the fetures will be enabled
 444     if (do_bind) {
 445         gms::get_local_gossiper().wait_for_gossip_to_settle().get();
 446     }
```

I agree waiting for nodes are marked as UP is a nice thing we should consider, but I do not want GOSSIP_SETTLE_MIN_WAIT_MS to be part of the solution.

What if we do not wait for nodes marked as up before we accept cql? Some of the writes might fail due to live replicas are not meet (in fact those nodes are alive). Anything else?

Reply:
> 
> What if we do not wait for nodes marked as up before we accept cql? Some of the writes might fail due to live replicas are not meet (in fact those nodes are alive). Anything else?

I can only think of failed requests (both reads and writes). It would be a regression of the following fix: https://issues.apache.org/jira/browse/CASSANDRA-4288

Reply:
@tgrabiec, GOSSIP_SETTLE_MIN_WAIT_MS, which today is set to 5 seconds, is one of the reasons why this patch set exists. On a reliable and fast network, there should be no reason to just sleep unconditionally for 5 seconds. If we are waiting for **something** to happen, we can can check it frequently instead of waiting 5 seconds before checking. That's the goal of this patch.